### ISSUE TITLE
Contextually type await expressions

### ISSUE THREAD

**Issue by omidkrad** [Microsoft/TypeScript#4904]
Since `await` expects a `Promise` type then with `noImplicitAny` option enabled the following should not give error that parameters `resolve` and `reject` have implicit any types.

```typescript
async f() {
    await (resolve, reject) => setTimeout(resolve, 0);
}
```

The compiler should know `resolve` and `reject` are functions, just like it knows in the following code:

```typescript
async g() {
    return new Promise((resolve, reject) => setTimeout(resolve, 0));
}
```

**Comment by mhegazy** [comment]
I believe the underlying issue is that the `await` expression is not contextually typed by the *appropriate* Promise constructor. if that happened, the noImplictAny errors should go away, and you would also get the completion support in your editor while typing the lambda.

**Comment by mhegazy** [comment]
updating the title to reflect the underlying issue.

**Comment by mhegazy** [comment]
//cc: @rbuckton

**Comment by omidkrad** [comment]
May be related: https://github.com/tc39/ecmascript-asyncawait/issues/64

**Comment by omidkrad** [comment]
I'm closing this since the proposed usage is not in the specs.

Related: https://github.com/tc39/ecmascript-asyncawait/issues/64
